src/VscPreviewExtension/src/extension.ts

```diff
@@ -0,0 +1,270 @@
+                documentUri = documentUri;
+                dfmProcess.callDfm();
+            }
+        });
+    });
+}
+
+// Check the file type
+function isMarkdownFile(document: TextDocument) {
+    // Prevent processing of own documents
+    return document.languageId === 'markdown' && document.uri.scheme !== 'markdown';
+}
+
+function getMarkdownUri(uri: Uri) {
+    return uri.with({ scheme: 'markdown', path: uri.path + '.rendered', query: uri.toString() });
+}
+
+function showPreview(dfmPreview: PreviewCore, uri?: Uri, sideBySide: boolean = false) {
+    if (window.activeTextEditor.document.languageId !== 'markdown') {
+        window.showErrorMessage("This is not a markdown file!");
+        return;
+    }
+    dfmPreview._isFirstTime = true;
+    let resource = uri;
+    if (!(resource instanceof Uri)) {
+        if (window.activeTextEditor) {
+            resource = window.activeTextEditor.document.uri;
+        }
+    }
+
+    if (!(resource instanceof Uri)) {
+        if (!window.activeTextEditor) {
+            // This is most likely toggling the preview
+            return commands.executeCommand('markdown.showSource');
+        }
+        // Nothing found that could be shown or toggled
+        return;
+    }
+
+    let thenable = commands.executeCommand('vscode.previewHtml',
+        getMarkdownUri(resource),
+        getViewColumn(sideBySide),
+        `Preview '${path.basename(resource.fsPath)}'`);
+
+    documentUri = getMarkdownUri(resource);
+    dfmPreview.callDfm();
+    return thenable;
+}
+
+function getViewColumn(sideBySide): ViewColumn {
+    const active = window.activeTextEditor;
+    if (!active) {
+        return ViewColumn.One;
+    }
+
+    if (!sideBySide) {
+        return active.viewColumn;
+    }
+
+    switch (active.viewColumn) {
+        case ViewColumn.One:
+            return ViewColumn.Two;
+        case ViewColumn.Two:
+            return ViewColumn.Three;
+    }
+
+    return active.viewColumn;
+}
+
+function showSource(mdUri: Uri) {
+    if (!mdUri) {
+        return commands.executeCommand('workbench.action.navigateBack');
+    }
+
+    const docUri = Uri.parse(mdUri.query);
+
+    for (let editor of window.visibleTextEditors) {
+        if (editor.document.uri.toString() === docUri.toString()) {
+            return window.showTextDocument(editor.document, editor.viewColumn);
+        }
+    }
+
+    return workspace.openTextDocument(docUri).then(doc => {
+        return window.showTextDocument(doc);
+    });
+}
+
+// Create a child process(DfmRender) by '_spawn' to render a html
+class PreviewCore {
+    private _spawn: child_process.ChildProcess;
+    private _waiting: boolean;
+    public _isFirstTime: boolean;
+
+    constructor(context: ExtensionContext) {
+        let extpath = context.asAbsolutePath('./DfmParse/PreviewCore.exe');
+        this._spawn = child_process.spawn(extpath);
+        this._waiting = false;
+
+        this._spawn.stdout.on('data', function (data) {
```

Comment:
>data [](start = 48, length = 4)

Can we add a parameter to avoid global variables? refer to: [https://github.com/dotnet/docfx/blob/v1.4/src/docfx.website.themes/default/ManagedReference.html.primary.js#L77](https://github.com/dotnet/docfx/blob/v1.4/src/docfx.website.themes/default/ManagedReference.html.primary.js#L77)